### ISSUE TITLE
Automatic plugin static files collection broken after install 

### ISSUE THREAD

**Issue by wolflu05** [inventree/InvenTree#7709]
### Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find a similar issue

### Describe the bug*

I have broken down the issues with plugin static files into multiple smaller ones:

### ~~1. Static files path~~ (fixed with #7763)

<details><summary>Details</summary>
<p>

If an `AppMixin` plugin before had a `static` folder and the django app was loaded, `invoke static` had already collected the static files in the root of the static folder so a file in `my-plugin/static/panel.js` was served by `/static/panel.js`.
*(to remove clashes between files with the same name from multiple plugins, I placed my static files in `my-plugin/static/my-plugin/panel.js`, so it was served by `/static/my-plugin/panel.js`)*

Now after the mentioned PR, a file in `my-plugin/static/panel.js` is served by `/static/plugins/my-plugin/panel.js`. I do not think this change is bad, it just needs to be documented as breaking. One good think would be to provide a `{% plugin_static "panel.js" %}` template tag which resolves to the particular plugin static url (`/static/plugins/my-plugin/panel.js`) and we back port that to 0.15 which resolves to `/static/panel.js`

</p>
</details> 

### 2. Automatic static file copying and migrations are not run after activating a plugin

I can reproduce this in a fresh docker setup (but not in my dev setup running locally). But no idea why only in a production docker setup. (Sometimes only very few times it actually succeeds, but then the migrations are run two times, the first time without any error, and the second they error out, because obviously the table already exists, that the migrations should create). Maybe that's some type of race condition.

0. Goto AdminCenter > Plugins > Settings and Activate plugins that uses apps
1. Goto AdminCenter > Plugins > Install plugin and use `inventree-bulk-plugin` (or `inventree-kicad`, but that only tests with migrations)
2. Activate that plugin
3. See this error in the docker logs: `inventree-worker  | Plugin registry has no record of plugin 'inventree-bulk-plugin'`
   - See that the folder `static/plugins/inventree-bulk-plugin` does not exist (folder should exist)
   - See that the migrantions have not run (should have run)
5. Run `invoke update` the files now exists and the migrations have run

### ~~3. When deactivating a plugin, the the static **files** don't get removed~~

When deactivating a plugin, the static files get don't removed automatically, a `invoke static` is needed. And even then, nested folders remain.

### ~~4. AppMixin Plugins with static files causes an error~~ (fixed with #7763)

<details><summary>Details</summary>
<p>

(If 2 is solved, found with my local dev setup where I don't have the issue 2, or when running `invoke static`)

1. Install a plugin with the app mixin that provides static files like in (2.1 and 2.2)
2. Activate the plugin
3. See the following error:
```
Process-c8d4a6053e4d446fb9f81077681a7693 processing blue-snake-xray-beryllium 'plugin.staticfiles.copy_plugin_static_files'
Copying static files for plugin 'inventree-bulk-plugin'
Failed 'plugin.staticfiles.copy_plugin_static_files' (blue-snake-xray-beryllium) - [Errno 66] Directory not empty: '/Users/wolflu/Development/1_GITHUB/InvenTree/dev/static/inventree-bulk-plugin' : Traceback (most recent call last):
  File "/Users/wolflu/Development/1_GITHUB/InvenTree/dev/venv/lib/python3.11/site-packages/django_q/worker.py", line 97, in worker
    res = f(*task["args"], **task["kwargs"])
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wolflu/Development/1_GITHUB/InvenTree/src/backend/InvenTree/plugin/staticfiles.py", line 66, in copy_plugin_static_files
    clear_static_dir(destination_prefix)
  File "/Users/wolflu/Development/1_GITHUB/InvenTree/src/backend/InvenTree/plugin/staticfiles.py", line 31, in clear_static_dir
    staticfiles_storage.delete(d)
  File "/Users/wolflu/Development/1_GITHUB/InvenTree/dev/venv/lib/python3.11/site-packages/django/core/files/storage/filesystem.py", line 156, in delete
    os.rmdir(name)
OSError: [Errno 66] Directory not empty: '/Users/wolflu/Development/1_GITHUB/InvenTree/dev/static/inventree-bulk-plugin'
```

</p>
</details> 

### ~~5. Property bug (can only sometimes reproduce)~~ => cannot reproduce anymore

When freshly installing a plugin (freshly, not previously installed), and then activating the plugin without restarting the server it can happen that it looks like this:
![image](https://github.com/user-attachments/assets/37778ae3-8584-47fd-9c86-d7b1ba3503b6)

<details><summary>Error</summary>
<p>

```
Traceback (most recent call last):

File "/root/.local/lib/python3.11/site-packages/rest_framework/views.py", line 506, in dispatch
response = handler(request, *args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/home/inventree/src/backend/InvenTree/plugin/api.py", line 341, in get
settings_definition=settings, plugin=plugin.plugin_config()
^^^^^^^^^^^^^^^^^^^^^^

TypeError: MetaBase.plugin_config() missing 1 required positional argument: 'self'
``` 

</p>
</details> 

### Deployment Method

- [X] Docker
- [ ] Package
- [ ] Bare metal
- [ ] Other - added info in Steps to Reproduce

### Version Information

```
# Version Information:
InvenTree-Version: 0.17.0 dev
Django Version: 4.2.15
Commit Hash: cc45c23
Commit Date: 2024-08-30

Database: postgresql
Debug-Mode: False
Deployed using Docker: True
Platform: Linux-6.6.31-linuxkit-x86_64-with
Installer: DOC

Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.1.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'InvenTreeCurrencyExchange', 'slug': 'inventreecurrencyexchange', 'version': '1.0.0'}, {'name': 'InvenTreeLabel', 'slug': 'inventreelabel', 'version': '1.1.0'}, {'name': 'InvenTreeLabelMachine', 'slug': 'inventreelabelmachine', 'version': '1.0.0'}, {'name': 'InvenTreeLabelSheet', 'slug': 'inventreelabelsheet', 'version': '1.0.0'}, {'name': 'DigiKeyPlugin', 'slug': 'digikeyplugin', 'version': '1.0.0'}, {'name': 'LCSCPlugin', 'slug': 'lcscplugin', 'version': '1.0.0'}, {'name': 'MouserPlugin', 'slug': 'mouserplugin', 'version': '1.0.0'}, {'name': 'TMEPlugin', 'slug': 'tmeplugin', 'version': '1.0.0'}]
```

### Please verify if you can reproduce this bug on the demo site.

- [ ] I can reproduce this bug on the demo site.

**Comment by SchrodingersGat** [comment]
@luwol03 a few different things to address here!

### Static File Pathing

- You say that the path for serving static files is different? This was not intentional. Can you provide examples of the differences?

### Static File Copying Does Not Work

- Let's work out why!

**Comment by wolflu05** [comment]
Good idea to break it into different issues:

### 1. Static files path
If an `AppMixin` plugin before had a `static` folder and the django app was loaded, `invoke static` had already collected the static files in the root of the static folder so a file in `my-plugin/static/panel.js` was served by `/static/panel.js`.
*(to remove clashes between files with the same name from multiple plugins, I placed my static files in `my-plugin/static/my-plugin/panel.js`, so it was served by `/static/my-plugin/panel.js`)*

Now after the mentioned PR, a file in `my-plugin/static/panel.js` is served by `/static/plugins/my-plugin/panel.js`. I do not think this change is bad, it just needs to be documented as breaking. One good think would be to provide a `{% plugin_static "panel.js" %}` template tag which resolves to the particular plugin static url (`/static/plugins/my-plugin/panel.js`) and we back port that to 0.15 which resolves to `/static/panel.js`

### 2. Automatic static file copying and migrations are not run after activating a plugin

I can repeatedly reproduce this in a fresh docker setup. But no idea why it's not woking.

1. Goto AdminCenter > Plugins > Install plugin and use `inventree-bulk-plugin`
2. Activate that plugin
3. See that the folder `static/plugins/inventree-bulk-plugin` does not exist (folder should exist)
4. Run `invoke static` the files now exists
5. Deactivate & Uninstall the plugin.
6. See that the files still exists (folder should not exists anymore)
7. Run `invoke static` again, the files are now removed (the nested folders remain)

### 3. When deactivating a plugin, the the static **files** don't removed 
When deactivating a plugin, the static files get don't removed automatically, a `invoke static` is needed. But that only removes the files, all folders are kept. Suggestion: use something like `shutil.rmtree(...)`, this removes a complete folder with all its files.

### 4. AppMixin Plugins with migrations, migrations do not run

1. Install the plugin like in (2.1 and 2.2) (or try the `inventree_kicad` plugin)
2. See that the migrations have not run by inspecting the database (there is no table `inventree_bulk_plugin_bulkcreationtemplate` or no entry in `django_migrations` with the app=inventree_bulk_plugin
3. Run `invoke migrate`
4. There where no migrations applied

### 5. Property bug

When freshly installing a plugin (freshly, not previously installed), and then activating the plugin without restarting the server it can happen that it looks like this:
![image](https://github.com/user-attachments/assets/37778ae3-8584-47fd-9c86-d7b1ba3503b6)

**Comment by matmair** [comment]
I think the migration part of 2 and 5 should be split out as individual issues as they do not seem connected with static content "stuff".

**Comment by wolflu05** [comment]
I agree with point 5, but when looking deeper, issue 2 targets exactly the code added in the static files for plugins pr. It works in dev, but not, (reproducable) in a fresh docker prod setup.

Can you reproduce that too in a fresh docker prod setup?

**Comment by SchrodingersGat** [comment]
@wolflu05 I think this has been addressed now, please let me know if this is incorrect...

**Comment by wolflu05** [comment]
Only 1 and 4 have been fixed be me in one of my recent prs. 2,3 and 5 have been still not fixed from here: https://github.com/inventree/InvenTree/issues/7709#issuecomment-2244844925

**Comment by SchrodingersGat** [comment]
@wolflu05 good point.

**Comment by wolflu05** [comment]
Sorry, but unfortunately I won't have any time the next weeks.

**Comment by matmair** [comment]
Is this issue still happening?

**Comment by wolflu05** [comment]
> Is this issue still happening?

Yes, I have updated the initial comment now with the list where I have broken down the issue into smaller ones (previously this was only a comment in this discussion, so that this doesn't get lost). Issue 2 and 3 are still not solved. And issue 5 is not reproducible. I have tested with the current latest docker image (updated the version information accordingly).

**Comment by wolflu05** [comment]
I have tested again with ec2051d, and 2,3 and 5 is still not resolved, so I'm reopening it.

Now removing the files after disabling the plugin is implemented, but the same error as 3 occurs when running without a restart. And I can actually reproduce 5 now again.

**Comment by SchrodingersGat** [comment]
@wolflu05 OK, let's work it out!

**Comment by SchrodingersGat** [comment]
> I have tested again with https://github.com/inventree/InvenTree/commit/ec2051d70a18bb8b82bdd9ff37b6bcf978acab54

@wolflu05 can you please describe further your steps to reproduce? This has been working well as far as I can tell, perhaps there are some differences between our setups

**Comment by wolflu05** [comment]
I have been using the default docker setup like described in the docs to reproduce it.

**Comment by SchrodingersGat** [comment]
Ok, I have worked out a fix for 2) and 3) - incoming

**Comment by SchrodingersGat** [comment]
@wolflu05 checkout https://github.com/inventree/InvenTree/pull/8502 and LMK what you think

**Comment by wolflu05** [comment]
@SchrodingersGat ok, I tested it now, it seems to be still not working correctly. I setup the system by using the default prod docker setup, just change the image to `latest` in the `.env` file. Then activated all plugin support (app mixin, urls, ...). Next installed `inventree-bulk-plugin`. And the I got the message `no plugin with key ... found`. I checked the db (if migrations from my AppMixin have been run) and static files, nothing there. I let the PC sit because I had something else todo, and when I came back, it seems like a job has picked up the migration work at least. For static files, I had to manually do a restart of the server+worker container.

But the static files removal works now pretty good after uninstalling the plugin (3). And I was not able to reproduce the property bug (5) anymore.
